### DISCUSSION TITLE
Run JanusGraph with Cassandra hosted in Kubernetes

### DISCUSSION THREAD

**Discussion by vmeln** [JanusGraph/janusgraph#2704] [Q&A] [answered]
Anyone tried to run JanusGraph with Cassandra hosted in Kubernetes? JanusGraph reads IP addresses [here and it seems that they persisted for the full lifetime of the graph](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLStoreManager.java#L258). 
Whenever the IP of the node change - the graph just crashes and not able to operate. My ideas are: 
- Push the traffic between JanusGraph  and Cassandra outside of the cluster through some load balancer
- Restart JanusGraph pods with readiness/liveness when Kubernetes move Cassandra pods

None of them are ideal or even production-worth. I hope someone can suggest something better.

**Comment by li-boxuan** [comment]
There is an issue: https://github.com/JanusGraph/janusgraph/issues/2464 maybe you could ask there and see if anyone has come up with a solution.

UPDATE: just noticed you have already seen that thread. I don't know if there is any workaround. Probably you could try fixing the source code and build JanusGraph by yourself.

**Reply by vmeln** [reply]
It's too big effort for me, at least I need to learn java and its ecosystem 😄
Stopped on placing Cassandra nodes in the different subnet. There are exceptions on JanusGraph initializing, but at least there seems to be no downtime.

**Comment by vmeln** [comment, marked as answer]
Small update. We used the K8ssandra operator to run the Cassandra cluster in Kubernetes and all issues described above during scale were caused by a replication factor of 1. We increased it to 3 and killing pods, scale - runs smoothly. Thread can be found here: https://github.com/k8ssandra/k8ssandra/issues/1045

**Reply by ryou90** [reply]
Hi :),
Sounds realy good. Do you have also solved the problem with the IP changes ? We are currently also plan to use Janusgraph in combination with k8ssndra. We are very interested in experiences with this setup. Do you use Cassandra 3 or 4 version?

Best,
Robert

**Reply by vmeln** [reply]
We use 3.11.10 version of Cassandra. [I'm not sure if 4th version is supported ](https://github.com/JanusGraph/janusgraph/releases/tag/v0.6.0). Seems using _*-dc1-service_ solves the issue with IP change. Anyways, in our case, it was just the default replication factor. However, we're going to run load tests, I'll update the thread(answer) if we discover something bad.

**Reply by ryou90** [reply]
Hi,
i have one question. How do you host Janusgraph in combination with K8ssandra? Run janusgraph on standalone cluster or as mixed pod setup (one janusgraph container and one cassandra container)?
Thx

**Reply by vmeln** [reply]
@ryou90 currently we run a single Janusgraph pod and standalone K8S cluster for Cassandra (3-4 instances). It depends on your security model or NFRs. You probably can host them alongside each other without issues.
